File: pingdartdb-dart/lib/src/query_builder.dart
import 'package:mysql1/mysql1.dart';
import 'dart:math';

class QueryBuilder {
  final String tableName;
  final MySqlConnection connection;
  final String dbType;
  final String q;

  QueryBuilder(this.tableName, this.connection, this.dbType)
      : q = (dbType == 'postgresql' || dbType == 'pgsql') ? '"' : '`';

  void _buildWhereClause(Map<String, dynamic> conditions, List<String> clauses, List<dynamic> replacements) {
    conditions.forEach((key, value) {
      if (value is Map) {
        if (value.containsKey('min') && value.containsKey('max')) {
          clauses.add("$q$key$q BETWEEN ? AND ?");
          replacements.addAll([value['min'], value['max']]);
        } else if (value.containsKey('start') && value.containsKey('end')) {
          clauses.add("$q$key$q BETWEEN ? AND ?");
          replacements.addAll([value['start'], value['end']]);
        }
      } else if (value is List) {
        if (value.isNotEmpty) {
          final placeholders = List.filled(value.length, '?').join(', ');
          clauses.add("$q$key$q IN ($placeholders)");
          replacements.addAll(value);
        } else {
          clauses.add("1=0");
        }
      } else if (value is String) {
        if (value.startsWith('!')) {
          clauses.add("$q$key$q != ?");
          replacements.add(value.substring(1));
        } else if (value.startsWith('>=')) {
          clauses.add("$q$key$q >= ?");
          replacements.add(value.substring(2));
        } else if (value.startsWith('<=')) {
          clauses.add("$q$key$q <= ?");
          replacements.add(value.substring(2));
        } else if (value.startsWith('>')) {
          clauses.add("$q$key$q > ?");
          replacements.add(value.substring(1));
        } else if (value.startsWith('<')) {
          clauses.add("$q$key$q < ?");
          replacements.add(value.substring(1));
        } else {
          clauses.add("$q$key$q = ?");
          replacements.add(value);
        }
      } else {
        clauses.add("$q$key$q = ?");
        replacements.add(value);
      }
    });
  }

  Future<void> _processMargedataBatch(List<Map<String, dynamic>> rows, Map<String, dynamic> margedataItem, Map<String, dynamic>? search) async {
    final targetTable = margedataItem['target_table']?.toString();
    final targetColumn = margedataItem['target_column']?.toString();
    final targetValue = margedataItem['target_value']?.toString();
    final targetLabel = margedataItem['target_label']?.toString();

    if (targetTable == null || targetColumn == null || targetValue == null || targetLabel == null) return;

    final parentIds = rows.map((row) => row[targetValue]).where((v) => v != null).toSet();

    if (parentIds.isEmpty) {
      for (var row in rows) { row[targetLabel] = []; }
      return;
    }

    final conditions = <String>[];
    final replacements = <dynamic>[];

    final placeholders = List.filled(parentIds.length, '?').join(', ');
    conditions.add("$q$targetColumn$q IN ($placeholders)");
    replacements.addAll(parentIds);

    final query = "SELECT * FROM $q$targetTable$q WHERE ${conditions.join(' AND ')}";

    try {
      final results = await connection.query(query, replacements);
      final allMatches = results.map((e) => Map<String, dynamic>.from(e.fields)).toList();

      final groupedMatches = <dynamic, List<Map<String, dynamic>>>{};
      for (var match in allMatches) {
        final pid = match[targetColumn];
        groupedMatches.putIfAbsent(pid, () => []).add(match);
      }

      for (var row in rows) {
        final pid = row[targetValue];
        row[targetLabel] = groupedMatches[pid] ?? [];
      }
    } catch (e) {
      for (var row in rows) { row[targetLabel] = []; }
    }
  }

  Future<Map<String, dynamic>> read([Map<String, dynamic>? options]) async {
    options ??= {};
    final conditions = options['conditions'] as Map<String, dynamic>? ?? {};
    final search = options['search'] as Map<String, dynamic>? ?? {};
    final orderBy = options['orderBy']?.toString();
    final pagination = options['pagination'] as Map<String, dynamic>?;
    final rangeData = options['range'] as Map<String, dynamic>?;
    final margedata = options['margedata'] as List<dynamic>? ?? [];

    var query = "SELECT * FROM $q$tableName$q WHERE 1=1";
    final replacements = <dynamic>[];

    final condClauses = <String>[];
    _buildWhereClause(conditions, condClauses, replacements);

    if (condClauses.isNotEmpty) {
      query += " AND ${condClauses.join(' AND ')}";
    }

    if (search.isNotEmpty) {
      final searchClauses = <String>[];
      search.forEach((key, value) {
        if (value != null && value.toString().isNotEmpty) {
          final likeOp = (dbType == 'postgresql') ? 'ILIKE' : 'LIKE';
          searchClauses.add("$q$key$q $likeOp ?");
          replacements.add("%$value%");
        }
      });
      if (searchClauses.isNotEmpty) {
        query += " AND (${searchClauses.join(' OR ')})";
      }
    }

    if (rangeData != null && rangeData.containsKey('latitude')) {
      final lat = rangeData['latitude'];
      final lng = rangeData['longitude'];
      final rad = rangeData['radius'];
      final tLat = rangeData['target_latitude'];
      final tLng = rangeData['target_longitude'];

      query += ''' AND (
        6371 * ACOS(
          COS(RADIANS(?)) *
          COS(RADIANS($q$tLat$q)) *
          COS(RADIANS($q$tLng$q) - RADIANS(?)) +
          SIN(RADIANS(?)) *
          SIN(RADIANS($q$tLat$q))
        )
      ) <= ?''';
      replacements.addAll([lat, lng, lat, rad]);
    }

    if (orderBy != null && orderBy.isNotEmpty) {
      query += " ORDER BY $orderBy";
    }

    final countQuery = query.replaceFirst('SELECT *', 'SELECT COUNT(*) as total');
    final countResult = await connection.query(countQuery, replacements);
    final totalCount = countResult.first[0];

    int totalPages = 0;
    if (pagination != null && pagination.containsKey('page') && pagination.containsKey('limit')) {
      final page = max(1, (pagination['page'] as num).toInt());
      final limit = (pagination['limit'] as num).toInt();
      final offset = (page - 1) * limit;
      query += " LIMIT $limit OFFSET $offset";
      totalPages = (totalCount / limit).ceil();
    }

    final results = await connection.query(query, replacements);
    final rows = results.map((e) => Map<String, dynamic>.from(e.fields)).toList();

    if (margedata.isNotEmpty && rows.isNotEmpty) {
      for (var mdItem in margedata) {
        if (mdItem is Map<String, dynamic>) {
          await _processMargedataBatch(rows, mdItem, search);
        }
      }
    }

    return {
      'success': true,
      'data': rows,
      'totalCount': totalCount,
      'totalPages': totalPages,
    };
  }

  Future<Map<String, dynamic>> update(Map<String, dynamic> data, Map<String, dynamic> conditions) async {
    if (conditions.isEmpty) throw Exception("Update requires conditions to prevent bulk overwrite.");

    final setClauses = <String>[];
    final replacements = <dynamic>[];

    data.forEach((k, v) {
      setClauses.add("$q$k$q = ?");
      replacements.add(v);
    });

    final condClauses = <String>[];
    _buildWhereClause(conditions, condClauses, replacements);

    if (condClauses.isEmpty) throw Exception("Update conditions could not be generated.");

    final query = "UPDATE $q$tableName$q SET ${setClauses.join(', ')} WHERE ${condClauses.join(' AND ')}";
    final result = await connection.query(query, replacements);

    return {
      'success': true,
      'message': 'Record(s) updated successfully.',
      'affectedRows': result.affectedRows,
    };
  }

  Future<Map<String, dynamic>> delete(Map<String, dynamic> conditions) async {
    if (conditions.isEmpty) throw Exception("Delete requires conditions to prevent bulk delete.");

    final condClauses = <String>[];
    final replacements = <dynamic>[];
    _buildWhereClause(conditions, condClauses, replacements);

    if (condClauses.isEmpty) throw Exception("Delete conditions could not be generated.");

    final query = "DELETE FROM $q$tableName$q WHERE ${condClauses.join(' AND ')}";
    final result = await connection.query(query, replacements);

    return {
      'success': true,
      'message': 'Record(s) deleted successfully.',
      'affectedRows': result.affectedRows,
    };
  }
}

File: pingdartdb-go/query_builder.go
package pingdartdb

import (
	"database/sql"
	"fmt"
	"math"
	"strings"
)

type QueryBuilder struct {
	TableName string
	DB        *sql.DB
	DbType    string
	Q         string
}

func NewQueryBuilder(tableName string, db *sql.DB, dbType string) *QueryBuilder {
	q := "`"
	if dbType == "postgresql" || dbType == "pgsql" {
		q = "\""
	}
	return &QueryBuilder{
		TableName: tableName,
		DB:        db,
		DbType:    dbType,
		Q:         q,
	}
}

func (qb *QueryBuilder) buildWhereClause(conditions map[string]interface{}) ([]string, []interface{}) {
	var clauses []string
	var replacements []interface{}

	for key, value := range conditions {
		switch v := value.(type) {
		case map[string]interface{}:
			if min, okMin := v["min"]; okMin {
				if max, okMax := v["max"]; okMax {
					clauses = append(clauses, fmt.Sprintf("%s%s%s BETWEEN ? AND ?", qb.Q, key, qb.Q))
					replacements = append(replacements, min, max)
				}
			} else if start, okStart := v["start"]; okStart {
				if end, okEnd := v["end"]; okEnd {
					clauses = append(clauses, fmt.Sprintf("%s%s%s BETWEEN ? AND ?", qb.Q, key, qb.Q))
					replacements = append(replacements, start, end)
				}
			}
		case []interface{}:
			if len(v) > 0 {
				placeholders := make([]string, len(v))
				for i := range v {
					placeholders[i] = "?"
					replacements = append(replacements, v[i])
				}
				clauses = append(clauses, fmt.Sprintf("%s%s%s IN (%s)", qb.Q, key, qb.Q, strings.Join(placeholders, ", ")))
			} else {
				clauses = append(clauses, "1=0")
			}
		case string:
			if strings.HasPrefix(v, "!") {
				clauses = append(clauses, fmt.Sprintf("%s%s%s != ?", qb.Q, key, qb.Q))
				replacements = append(replacements, v[1:])
			} else if strings.HasPrefix(v, ">=") {
				clauses = append(clauses, fmt.Sprintf("%s%s%s >= ?", qb.Q, key, qb.Q))
				replacements = append(replacements, v[2:])
			} else if strings.HasPrefix(v, "<=") {
				clauses = append(clauses, fmt.Sprintf("%s%s%s <= ?", qb.Q, key, qb.Q))
				replacements = append(replacements, v[2:])
			} else if strings.HasPrefix(v, ">") {
				clauses = append(clauses, fmt.Sprintf("%s%s%s > ?", qb.Q, key, qb.Q))
				replacements = append(replacements, v[1:])
			} else if strings.HasPrefix(v, "<") {
				clauses = append(clauses, fmt.Sprintf("%s%s%s < ?", qb.Q, key, qb.Q))
				replacements = append(replacements, v[1:])
			} else {
				clauses = append(clauses, fmt.Sprintf("%s%s%s = ?", qb.Q, key, qb.Q))
				replacements = append(replacements, v)
			}
		default:
			clauses = append(clauses, fmt.Sprintf("%s%s%s = ?", qb.Q, key, qb.Q))
			replacements = append(replacements, v)
		}
	}

	return clauses, replacements
}

// Executes arbitrary query and returns list of generic maps
func (qb *QueryBuilder) fetchRows(query string, args []interface{}) ([]map[string]interface{}, error) {
	rows, err := qb.DB.Query(query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	cols, err := rows.Columns()
	if err != nil {
		return nil, err
	}

	var results []map[string]interface{}
	for rows.Next() {
		columns := make([]interface{}, len(cols))
		columnPointers := make([]interface{}, len(cols))
		for i := range columns {
			columnPointers[i] = &columns[i]
		}

		if err := rows.Scan(columnPointers...); err != nil {
			return nil, err
		}

		rowMap := make(map[string]interface{})
		for i, colName := range cols {
			val := columnPointers[i].(*interface{})
			if b, ok := (*val).([]byte); ok {
				rowMap[colName] = string(b)
			} else {
				rowMap[colName] = *val
			}
		}
		results = append(results, rowMap)
	}

	return results, nil
}

func (qb *QueryBuilder) processMargedata(rows []map[string]interface{}, margedataItem map[string]interface{}, search map[string]interface{}) {
	targetTable, _ := margedataItem["target_table"].(string)
	targetColumn, _ := margedataItem["target_column"].(string)
	targetValue, _ := margedataItem["target_value"].(string)
	targetLabel, _ := margedataItem["target_label"].(string)

	if targetTable == "" || targetColumn == "" || targetValue == "" || targetLabel == "" {
		return
	}

	parentIdsMap := make(map[interface{}]bool)
	for _, row := range rows {
		if val, ok := row[targetValue]; ok && val != nil {
			parentIdsMap[val] = true
		}
	}

	if len(parentIdsMap) == 0 {
		for _, row := range rows {
			row[targetLabel] = []map[string]interface{}{}
		}
		return
	}

	var conditions []string
	var replacements []interface{}

	var placeholders []string
	for pid := range parentIdsMap {
		placeholders = append(placeholders, "?")
		replacements = append(replacements, pid)
	}
	conditions = append(conditions, fmt.Sprintf("%s%s%s IN (%s)", qb.Q, targetColumn, qb.Q, strings.Join(placeholders, ", ")))

	query := fmt.Sprintf("SELECT * FROM %s%s%s WHERE %s", qb.Q, targetTable, qb.Q, strings.Join(conditions, " AND "))
	allMatches, err := qb.fetchRows(query, replacements)
	if err != nil {
		for _, row := range rows {
			row[targetLabel] = []map[string]interface{}{}
		}
		return
	}

	groupedMatches := make(map[interface{}][]map[string]interface{})
	for _, match := range allMatches {
		pid := match[targetColumn]
		// Handle string parsing issue if numeric ID was read as string byte
		if b, ok := pid.([]byte); ok {
			pid = string(b)
		}
		
		groupedMatches[pid] = append(groupedMatches[pid], match)
	}

	for _, row := range rows {
		pid := row[targetValue]
		if b, ok := pid.([]byte); ok {
			pid = string(b)
		}
		
		if matches, ok := groupedMatches[pid]; ok {
			row[targetLabel] = matches
		} else {
			row[targetLabel] = []map[string]interface{}{}
		}
	}
}

func (qb *QueryBuilder) Read(options map[string]interface{}) (map[string]interface{}, error) {
	if options == nil {
		options = make(map[string]interface{})
	}

	conditions, _ := options["conditions"].(map[string]interface{})
	search, _ := options["search"].(map[string]interface{})
	orderBy, _ := options["orderBy"].(string)
	pagination, _ := options["pagination"].(map[string]interface{})
	rangeData, _ := options["range"].(map[string]interface{})
	margedata, _ := options["margedata"].([]interface{})

	query := fmt.Sprintf("SELECT * FROM %s%s%s WHERE 1=1", qb.Q, qb.TableName, qb.Q)
	var replacements []interface{}

	condClauses, condReps := qb.buildWhereClause(conditions)
	if len(condClauses) > 0 {
		query += " AND " + strings.Join(condClauses, " AND ")
		replacements = append(replacements, condReps...)
	}

	if len(search) > 0 {
		var searchClauses []string
		for key, value := range search {
			if strVal, ok := value.(string); ok && strVal != "" {
				likeOp := "LIKE"
				if qb.DbType == "postgresql" {
					likeOp = "ILIKE"
				}
				searchClauses = append(searchClauses, fmt.Sprintf("%s%s%s %s ?", qb.Q, key, qb.Q, likeOp))
				replacements = append(replacements, "%"+strVal+"%")
			}
		}
		if len(searchClauses) > 0 {
			query += " AND (" + strings.Join(searchClauses, " OR ") + ")"
		}
	}

	if rangeData != nil {
		if lat, okLat := rangeData["latitude"].(float64); okLat {
			if lng, okLng := rangeData["longitude"].(float64); okLng {
				if rad, okRad := rangeData["radius"].(float64); okRad {
					tLat, _ := rangeData["target_latitude"].(string)
					tLng, _ := rangeData["target_longitude"].(string)

					query += fmt.Sprintf(` AND (
						6371 * ACOS(
							COS(RADIANS(?)) *
							COS(RADIANS(%s%s%s)) *
							COS(RADIANS(%s%s%s) - RADIANS(?)) +
							SIN(RADIANS(?)) *
							SIN(RADIANS(%s%s%s))
						)
					) <= ?`, qb.Q, tLat, qb.Q, qb.Q, tLng, qb.Q, qb.Q, tLat, qb.Q)
					replacements = append(replacements, lat, lng, lat, rad)
				}
			}
		}
	}

	if orderBy != "" {
		query += " ORDER BY " + orderBy
	}

	countQuery := strings.Replace(query, "SELECT *", "SELECT COUNT(*) as total", 1)
	var totalCount int
	err := qb.DB.QueryRow(countQuery, replacements...).Scan(&totalCount)
	if err != nil && err != sql.ErrNoRows {
		return nil, err
	}

	var totalPages float64 = 0
	if pagination != nil {
		pageFl, _ := pagination["page"].(float64)
		limitFl, _ := pagination["limit"].(float64)
		
		page := int(pageFl)
		if page < 1 {
			page = 1
		}
		limit := int(limitFl)
		
		if limit > 0 {
			offset := (page - 1) * limit
			query += fmt.Sprintf(" LIMIT %d OFFSET %d", limit, offset)
			totalPages = math.Ceil(float64(totalCount) / float64(limit))
		}
	}

	results, err := qb.fetchRows(query, replacements)
	if err != nil {
		return nil, err
	}

	if len(margedata) > 0 && len(results) > 0 {
		for _, mdItem := range margedata {
			if mdMap, ok := mdItem.(map[string]interface{}); ok {
				qb.processMargedata(results, mdMap, search)
			}
		}
	}

	return map[string]interface{}{
		"success":    true,
		"data":       results,
		"totalCount": totalCount,
		"totalPages": totalPages,
	}, nil
}

func (qb *QueryBuilder) Delete(conditions map[string]interface{}) (map[string]interface{}, error) {
	if len(conditions) == 0 {
		return nil, fmt.Errorf("Delete requires conditions to prevent bulk delete.")
	}

	condClauses, replacements := qb.buildWhereClause(conditions)
	if len(condClauses) == 0 {
		return nil, fmt.Errorf("Delete conditions could not be generated.")
	}

	query := fmt.Sprintf("DELETE FROM %s%s%s WHERE %s", qb.Q, qb.TableName, qb.Q, strings.Join(condClauses, " AND "))
	result, err := qb.DB.Exec(query, replacements...)
	if err != nil {
		return nil, err
	}

	affected, _ := result.RowsAffected()
	return map[string]interface{}{
		"success":      true,
		"message":      "Record(s) deleted successfully.",
		"affectedRows": affected,
	}, nil
}

func (qb *QueryBuilder) Update(data map[string]interface{}, conditions map[string]interface{}) (map[string]interface{}, error) {
	if len(conditions) == 0 {
		return nil, fmt.Errorf("Update requires conditions to prevent bulk overwrite.")
	}

	var setClauses []string
	var replacements []interface{}

	for k, v := range data {
		setClauses = append(setClauses, fmt.Sprintf("%s%s%s = ?", qb.Q, k, qb.Q))
		replacements = append(replacements, v)
	}

	condClauses, condReps := qb.buildWhereClause(conditions)
	if len(condClauses) == 0 {
		return nil, fmt.Errorf("Update conditions could not be generated.")
	}
	replacements = append(replacements, condReps...)

	query := fmt.Sprintf("UPDATE %s%s%s SET %s WHERE %s", qb.Q, qb.TableName, qb.Q, strings.Join(setClauses, ", "), strings.Join(condClauses, " AND "))
	
	result, err := qb.DB.Exec(query, replacements...)
	if err != nil {
		return nil, err
	}

	affected, _ := result.RowsAffected()
	return map[string]interface{}{
		"success":      true,
		"message":      "Record(s) updated successfully.",
		"affectedRows": affected,
	}, nil
}
